### ISSUE TITLE
bug: Call to getPageSource failed when call is returning a 200 (Android simulator)

### ISSUE THREAD

**Issue by zach-fetch** [appium/appium-inspector#1729]
### Is this an issue specifically with Appium Inspector?

- [X] I have verified that my issue does _not_ occur with Appium, and should be investigated as an Appium Inspector issue

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

When I refresh the source and screenshot on our app I get an error that says Call to `'getPageSource' failed
Unicode replacement character detected, source encoding issues?`. This doesn't make sense because the getPageSource is returning a 200.

This is only an issue with Android and not an issue in iOS.

Appium Logging:
```
[HTTP] --> GET /wd/hub/session/2f9c6893-52a1-4a96-a9bf-c64400ff5a14/source {}
[AndroidUiautomator2Driver@7c85] Driver proxy active, passing request on via HTTP proxy
[AndroidUiautomator2Driver@7c85] Matched '/wd/hub/session/2f9c6893-52a1-4a96-a9bf-c64400ff5a14/source' to command name 'getPageSource'
[AndroidUiautomator2Driver@7c85] Proxying [GET /wd/hub/session/2f9c6893-52a1-4a96-a9bf-c64400ff5a14/source] to [GET http://127.0.0.1:8200/session/269aad1b-70be-4934-bc62-67428cca8392/source] with no body
[AndroidUiautomator2Driver@7c85] Got response with status 200: {"sessionId":"269aad1b-70be-4934-bc62-67428cca8392","value":"<?xml version='1.0' encoding='UTF-8' standalone='yes' ?>\r\n<hierarchy index=\"0\" class=\"hierarchy\" rotation=\"0\" width=\"1440\" height=\"2891\">\r\n  <android.widget.FrameLayout index=\"0\" package=\"com.fetchrewards.fetchrewards.hop.debug\" class=\"android.widget.FrameLayout\" text=\"\" checkable=\"false\" checked=\"false\" clickable=\"false\" enabled=\"true\" focusable=\"false\" focused=\"false\" long-clickable=\"false\" password=\"false\" scrollable=\"false\" selected=\"false\" bounds=\"[0,0][1440,3120]\" displayed=\"true\">\r\n    <android.widget.LinearLayout index=\"0\" package=\"com.fetchrewards.fetchrewards.hop.debug\" class=\"android.widget.LinearLayout\" text=\"\" checkable=\"false\" checked=\"false\" clickable=\"false\" enabled=\"true\" focusable=\"false\" focused=\"false\" long-clickable=\"false\" password=\"false\" scrollable=\"false\" selected=\"false\" bounds=\"[0,0][1440,3120]\" displayed=\"true\">\r\n      <android.widget.Fra...
[AndroidUiautomator2Driver@7c85] Replacing sessionId 269aad1b-70be-4934-bc62-67428cca8392 with 2f9c6893-52a1-4a96-a9bf-c64400ff5a14
[HTTP] <-- GET /wd/hub/session/2f9c6893-52a1-4a96-a9bf-c64400ff5a14/source 200 76 ms - 75568
```

### Expected Behavior

The getPageSource call should be processed and the app page should appear in appium inspector

### Operating System

Mac

### Appium Inspector Version

2024.9.1 (not an issue in appium 2024.3.4)

### Appium Version

2.6.0

### Further Information

Driver version
- uiautomator2@3.5.2 [installed (npm)]

Simulator = Pixel 6 pro running Android 13

**Comment by eglitise** [comment]
Could you please provide the full Appium log?
Would it also be possible for you to check the other Inspector versions between these two, to identify the first version with this issue?

**Comment by eglitise** [comment]
Closing due to no response

**Comment by AlikShegidevich** [comment]
Hi! The issue has been reproduced on Mac + Simulator Pixel 8 Pro Android 14:
Appium Inspector 2024.9.1
Previous version 2024.8.2 doesn't have the issue.
Log and screenshot are attached. Not all screens of our android app affected by the issue. Some of them can be inspected, some not. The log covers multiple attempts to Refresh different screens. Last two attempts failed.

uiautomator2@3.5.2 [installed (npm)]
Appium 2.6.0

![Screenshot 2024-11-01 at 4 29 40 PM](https://github.com/user-attachments/assets/8fe1ea2c-992a-4048-8840-f83af6bda79e)

[appium.log](https://github.com/user-attachments/files/17604041/appium.log)

**Comment by AlikShegidevich** [comment]
@eglitise Please reopen the bug - I've submitted the requested info

**Comment by eglitise** [comment]
@AlikShegidevich thanks for the info. Could you also try to reproduce this using https://inspector.appiumpro.com/ ? I would like to check whether the problem is specific to Electron, or affects both versions.

**Comment by mykola-mokhnach** [comment]
@eglitise The below lines look highly suspicious to me:

> [HTTP] <-- GET /wd/hub/session/f5cdc235-2222-41e1-bba2-3b75c455ab8f/source 304 367 ms - -
> [HTTP] <-- GET /wd/hub/session/f5cdc235-2222-41e1-bba2-3b75c455ab8f/timeouts 304 135 ms - -

This looks like inspector is caching server responses, while it never should.

**Comment by eglitise** [comment]
Closing due to no response

**Comment by sreejits139** [comment]
I am still seeing the issue and not sure how to proceed. Can anyone please let me know if you found a solution for this? The issue happens on https://inspector.appiumpro.com/ too.
<img width="389" alt="image" src="https://github.com/user-attachments/assets/829ff0a0-184f-462a-a8ed-cea3b8893483">

**Comment by mykola-mokhnach** [comment]
It might be a duplicate of https://github.com/appium/appium/issues/20503#issuecomment-2357928628

**Comment by sreejits139** [comment]
I tried to compare but not sure if they are the same. I tried every option but somehow it errors out all the time say maybe I get lucky once in every 50 attempts!! Can you please let me know any alternative maybe to continue working.. Thanks for the help.

**Comment by mykola-mokhnach** [comment]
> I tried every option

What exactly did you try?

**Comment by sreejits139** [comment]
Node version: 20.61. - As most stable with webdriverio and appium. Started an emulator, started appium server, open the appium inspector and connect to an Android APK. After navigating to some pages that are like overlay pages, when I try to refresh the session on the inspector, it loads the 'Call to getPageSource failed' as above. Do you think I should try a newer version of Node? I had 23 but it was causing issues with webdriverio and appium so downgraded recently.

**Comment by mykola-mokhnach** [comment]
Try version 22.11.0

**Comment by sreejits139** [comment]
I installed the node 22.11.0, but I still get the same error. Can it be because of the app page itself as other pages are fine most of the times? Here is what I am doing:
1. Start a local appium server.
2. Start a local emulator in Android studio.
3. Open Appium inspector.
4. Once the app opens up, I click on the user icon on load a Sign in screen/overlay type of screen.
5. Now, I need to find the selector for the Join element. So, I click on the refresh button on the inspector and I see the same error.
![Screenshot 2024-11-20 at 9 12 48 PM](https://github.com/user-attachments/assets/827d1234-dfc6-427a-9ddb-ddeb10dd2d57)
Thanks again for the support!!!

**Comment by eglitise** [comment]
@sreejits139 is this app publicly available? I would like to reproduce this on my end.

**Comment by eglitise** [comment]
My current guess is that this might be caused by the `@xmldom/xmldom` upgrade from `0.8.10` to `0.9.3`.

**Comment by eglitise** [comment]
Ok, I think I can reproduce this by adding [the unicode replacement character](https://en.wikipedia.org/wiki/Specials_(Unicode_block)#Replacement_character) in the application DOM. With xmldom 0.8.10, it is being parsed without any issues.

**Comment by eglitise** [comment]
The issue should be fixed in version 2024.12.1. Thanks to everyone who reported it!
